File: src/drivers/duckdb/schema.ts
import type {
  TableInfo,
  ColumnInfo,
  IndexInfo,
  ConstraintInfo,
} from "../types";
import { execDuckDbJson } from "./subprocess";

type SchemaOpts = {
  dbPath: string | undefined;
  readonly: boolean;
};

const runQuery = async (
  opts: SchemaOpts,
  sql: string,
): Promise<Record<string, unknown>[]> =>
  execDuckDbJson({ dbPath: opts.dbPath, sql, readonly: opts.readonly });

export const createDuckDbSchema = (opts: SchemaOpts) => {
  const getTables = async (
    tableOpts?: { includeSystem?: boolean },
  ): Promise<TableInfo[]> => {
    const rows = await runQuery(
      opts,
      tableOpts?.includeSystem
        ? "SELECT table_name, table_type FROM information_schema.tables ORDER BY table_name"
        : "SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = 'main' ORDER BY table_name",
    );
    return rows.map((r) => ({
      name: r.table_name as string,
      type:
        (r.table_type as string) === "VIEW"
          ? ("view" as const)
          : ("table" as const),
    }));
  };

  const describeTable = async (table: string): Promise<ColumnInfo[]> => {
    const escaped = table.replace(/'/g, "''");
    const rows = await runQuery(
      opts,
      `SELECT column_name, data_type, is_nullable, column_default FROM information_schema.columns WHERE table_schema = 'main' AND table_name = '${escaped}' ORDER BY ordinal_position`,
    );
    return rows.map((r) => ({
      name: r.column_name as string,
      type: r.data_type as string,
      nullable: (r.is_nullable as string) === "YES",
      defaultValue: (r.column_default as string) ?? undefined,
    }));
  };

  const getIndexes = async (table?: string): Promise<IndexInfo[]> => {
    const where = table
      ? ` WHERE table_name = '${table.replace(/'/g, "''")}'`
      : "";
    const rows = await runQuery(
      opts,
      `SELECT index_name, table_name, is_unique, expressions FROM duckdb_indexes()${where} ORDER BY index_name`,
    );

    return rows.map((r) => ({
      name: r.index_name as string,
      table: r.table_name as string,
      columns: parseExpressionList(r.expressions as string),
      unique: (r.is_unique as string) === "true",
    }));
  };

  const getConstraints = async (
    table?: string,
  ): Promise<ConstraintInfo[]> => {
    const where = table
      ? ` WHERE table_name = '${table.replace(/'/g, "''")}'`
      : "";
    const rows = await runQuery(
      opts,
      `SELECT constraint_type, table_name, constraint_column_names, constraint_column_indexes FROM duckdb_constraints()${where} ORDER BY table_name`,
    );

    return rows
      .filter((r) => mapConstraintType(r.constraint_type as string))
      .map((r) => ({
        name: `${r.table_name}_${mapConstraintType(r.constraint_type as string)}`,
        table: r.table_name as string,
        type: mapConstraintType(r.constraint_type as string)!,
        columns: parseColumnList(r.constraint_column_names),
      }));
  };

  const searchSchema = async (
    pattern: string,
  ): Promise<{
    tables: TableInfo[];
    columns: { table: string; column: string }[];
  }> => {
    const escaped = pattern.replace(/'/g, "''");
    const [tableRows, colRows] = await Promise.all([
      runQuery(
        opts,
        `SELECT table_name FROM information_schema.tables WHERE table_schema = 'main' AND table_name LIKE '%${escaped}%' ORDER BY table_name`,
      ),
      runQuery(
        opts,
        `SELECT table_name, column_name FROM information_schema.columns WHERE table_schema = 'main' AND column_name LIKE '%${escaped}%' ORDER BY table_name, column_name`,
      ),
    ]);

    return {
      tables: tableRows.map((r) => ({ name: r.table_name as string })),
      columns: colRows.map((r) => ({
        table: r.table_name as string,
        column: r.column_name as string,
      })),
    };
  };

  return { getTables, describeTable, getIndexes, getConstraints, searchSchema };
};

const mapConstraintType = (
  duckType: string,
): "primary_key" | "foreign_key" | "unique" | "check" | undefined => {
  const map: Record<string, "primary_key" | "foreign_key" | "unique" | "check"> = {
    "PRIMARY KEY": "primary_key",
    "FOREIGN KEY": "foreign_key",
    UNIQUE: "unique",
    CHECK: "check",
  };
  return map[duckType];
};

const parseColumnList = (value: unknown): string[] => {
  if (Array.isArray(value)) {
    return value.map(String);
  }
  if (typeof value === "string") {
    return value.replace(/^\[|\]$/g, "").split(",").map((s) => s.trim());
  }
  return [];
};

// DuckDB expressions field format: "[col1, col2]"
const parseExpressionList = (expr: string): string[] => {
  if (!expr) {
    return [];
  }
  return expr
    .replace(/^\[|\]$/g, "")
    .split(",")
    .map((c) => c.trim())
    .filter(Boolean);
};

File: internal/driver/duckdb/schema.go
package duckdb

import (
	"context"
	"fmt"
	"strings"

	"github.com/shhac/agent-sql/internal/driver"
)

func (c *duckdbConn) GetTables(ctx context.Context, includeSystem bool) ([]driver.TableInfo, error) {
	query := "SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = 'main' ORDER BY table_name"
	if includeSystem {
		query = "SELECT table_name, table_type FROM information_schema.tables ORDER BY table_name"
	}

	rows, err := c.exec(ctx, query)
	if err != nil {
		return nil, err
	}

	tables := make([]driver.TableInfo, 0, len(rows))
	for _, r := range rows {
		name, _ := r["table_name"].(string)
		tableType, _ := r["table_type"].(string)
		t := "table"
		if tableType == "VIEW" {
			t = "view"
		}
		tables = append(tables, driver.TableInfo{Name: name, Type: t})
	}
	return tables, nil
}

func (c *duckdbConn) DescribeTable(ctx context.Context, table string) ([]driver.ColumnInfo, error) {
	escaped := strings.ReplaceAll(table, "'", "''")
	query := fmt.Sprintf(
		"SELECT column_name, data_type, is_nullable, column_default FROM information_schema.columns WHERE table_schema = 'main' AND table_name = '%s' ORDER BY ordinal_position",
		escaped,
	)

	rows, err := c.exec(ctx, query)
	if err != nil {
		return nil, err
	}

	columns := make([]driver.ColumnInfo, 0, len(rows))
	for _, r := range rows {
		name, _ := r["column_name"].(string)
		dataType, _ := r["data_type"].(string)
		nullable, _ := r["is_nullable"].(string)
		col := driver.ColumnInfo{
			Name:     name,
			Type:     dataType,
			Nullable: nullable == "YES",
		}
		if dflt, ok := r["column_default"]; ok && dflt != nil {
			col.DefaultValue = fmt.Sprintf("%v", dflt)
		}
		columns = append(columns, col)
	}
	return columns, nil
}

func (c *duckdbConn) GetIndexes(ctx context.Context, table string) ([]driver.IndexInfo, error) {
	query := "SELECT index_name, table_name, is_unique, expressions FROM duckdb_indexes() ORDER BY index_name"
	if table != "" {
		escaped := strings.ReplaceAll(table, "'", "''")
		query = fmt.Sprintf(
			"SELECT index_name, table_name, is_unique, expressions FROM duckdb_indexes() WHERE table_name = '%s' ORDER BY index_name",
			escaped,
		)
	}

	rows, err := c.exec(ctx, query)
	if err != nil {
		return nil, err
	}

	indexes := make([]driver.IndexInfo, 0, len(rows))
	for _, r := range rows {
		name, _ := r["index_name"].(string)
		tbl, _ := r["table_name"].(string)
		expr, _ := r["expressions"].(string)
		unique := toBool(r["is_unique"])
		indexes = append(indexes, driver.IndexInfo{
			Name:    name,
			Table:   tbl,
			Columns: parseExpressionList(expr),
			Unique:  unique,
		})
	}
	return indexes, nil
}

func (c *duckdbConn) GetConstraints(ctx context.Context, table string) ([]driver.ConstraintInfo, error) {
	query := "SELECT constraint_type, table_name, constraint_column_names FROM duckdb_constraints() ORDER BY table_name"
	if table != "" {
		escaped := strings.ReplaceAll(table, "'", "''")
		query = fmt.Sprintf(
			"SELECT constraint_type, table_name, constraint_column_names FROM duckdb_constraints() WHERE table_name = '%s' ORDER BY table_name",
			escaped,
		)
	}

	rows, err := c.exec(ctx, query)
	if err != nil {
		return nil, err
	}

	var constraints []driver.ConstraintInfo
	for _, r := range rows {
		constraintType, _ := r["constraint_type"].(string)
		mapped := driver.MapConstraintType(constraintType)
		if mapped == "" {
			continue
		}
		tbl, _ := r["table_name"].(string)
		constraints = append(constraints, driver.ConstraintInfo{
			Name:    fmt.Sprintf("%s_%s", tbl, mapped),
			Table:   tbl,
			Type:    mapped,
			Columns: parseColumnList(r["constraint_column_names"]),
		})
	}
	return constraints, nil
}

func (c *duckdbConn) SearchSchema(ctx context.Context, pattern string) (*driver.SearchResult, error) {
	escaped := strings.ReplaceAll(pattern, "'", "''")

	tableQuery := fmt.Sprintf(
		"SELECT table_name FROM information_schema.tables WHERE table_schema = 'main' AND table_name LIKE '%%%s%%' ORDER BY table_name",
		escaped,
	)
	colQuery := fmt.Sprintf(
		"SELECT table_name, column_name FROM information_schema.columns WHERE table_schema = 'main' AND column_name LIKE '%%%s%%' ORDER BY table_name, column_name",
		escaped,
	)

	tableRows, err := c.exec(ctx, tableQuery)
	if err != nil {
		return nil, err
	}

	colRows, err := c.exec(ctx, colQuery)
	if err != nil {
		return nil, err
	}

	tables := make([]driver.TableInfo, 0, len(tableRows))
	for _, r := range tableRows {
		name, _ := r["table_name"].(string)
		tables = append(tables, driver.TableInfo{Name: name})
	}

	columns := make([]driver.ColumnMatch, 0, len(colRows))
	for _, r := range colRows {
		tbl, _ := r["table_name"].(string)
		col, _ := r["column_name"].(string)
		columns = append(columns, driver.ColumnMatch{Table: tbl, Column: col})
	}

	return &driver.SearchResult{Tables: tables, Columns: columns}, nil
}

// --- shared helpers used by schema methods ---

// parseExpressionList parses DuckDB's expressions format: "[col1, col2]".
func parseExpressionList(expr string) []string {
	if expr == "" {
		return nil
	}
	expr = strings.TrimPrefix(expr, "[")
	expr = strings.TrimSuffix(expr, "]")
	parts := strings.Split(expr, ",")
	result := make([]string, 0, len(parts))
	for _, p := range parts {
		trimmed := strings.TrimSpace(p)
		if trimmed != "" {
			result = append(result, trimmed)
		}
	}
	return result
}

// parseColumnList parses DuckDB constraint_column_names which may be a
// JSON array or a bracket string.
func parseColumnList(value any) []string {
	switch v := value.(type) {
	case []any:
		result := make([]string, 0, len(v))
		for _, item := range v {
			result = append(result, fmt.Sprintf("%v", item))
		}
		return result
	case string:
		return parseExpressionList(v)
	default:
		return nil
	}
}

func toBool(v any) bool {
	switch b := v.(type) {
	case bool:
		return b
	case string:
		return b == "true"
	default:
		return false
	}
}
